interviews/elixir/server/lib/server/number_storage.ex

```diff
@@ -0,0 +1,30 @@
+# Global storage used to store and update the array submitted by the user
+defmodule Server.NumberStorage do
+  use Agent
+
+  def start_link() do
+    Agent.start(fn -> [] end, name: NumberStorage)
+  end
+
+  def add_list(new_list) do
+    Agent.update(NumberStorage, fn list -> new_list end)
+  end
+
+  def get_list() do
+    Agent.get(NumberStorage, fn list -> list end)
+  end
+
+  def add_num(num) do
+    list = Agent.get(NumberStorage, fn list -> list end)
```

Comment:
Interesting, you call Agent at the beginning and the end, with computation (using Elixir's super cool piping operator of course) in between.
Assume we have a very popular application with thousands of requests to this API per second and we have two requests that happen in immediate succession within the same millisecond.  Will both requests be viewing the same list when they call Agent.get(..)? And if so, does that mean that the second request will be adding its number to a list and updating the latest list that will not include the number from the first request? (Because the second request perhaps retrieved the list before the first request had completed updating it before it was retrieved by the first list).
Is this a problem here or is it solved somehow by the framework?  If it is a problem, how would you solve it?

Reply:
Hey Charles,

You are absolutely right that the two consecutive client calls would be modifying the same list. I totally missed that. All Agents are separate processes that manage state and rely on a message queue in order to determine when to perform certain actions(i.e. updates). As soon as I returned the list via the first `Agent.get(...)` function, I introduced the issue in which the two requests were able to perform different modifications to the same global state. The last request to finish transforming the list would end up overwriting the changes of the previous request.

In order to fix this, I should have put all of my list modifications in the `Agent.update(...)` instead. 'Update' requires a reply to be returned to the calling process in order to resume operations. So by putting all of the list transformations inside of the update function, I can utilize the Agents message queue to provide some sort of synchronicity to the list operations:

```elixir
Agent.update(NumberStorage, fn list ->
   |> Enum.split_while(fn el -> el < num end)
   |> Tuple.to_list() 
   |> List.insert_at(1, [num])
   |> List.flatten()
   |>IO.inspect()
end
```

Reply:
:+1: